drivers/src/ecc384.rs

```diff
@@ -289,6 +293,34 @@ impl Ecc384 {
         digest: &Ecc384Scalar,
         signature: &Ecc384Signature,
     ) -> CaliptraResult<bool> {
+        // Get the verify r result
+        let mut verify_r = self.verify_r(pub_key, digest, signature)?;
+
+        // compare the hardware generate `r` with one in signature
+        let result = verify_r == signature.r;
+        verify_r.0.fill(0);
+        Ok(result)
+    }
+
+    /// Returns the R value of the signature with specified public key and digest.
+    ///  Caller is expected to compare the returned R value against the provided signature's
+    ///  R value to determine whether the signature is valid.
+    ///
+    /// # Arguments
+    ///
+    /// * `pub_key` - Public key
+    /// * `digest` - digest to verify
+    /// * `signature` - Signature to verify
+    ///
+    /// # Result
+    ///
+    /// *  `Array4xN<12, 48>` - verify R value
+    pub fn verify_r(
```

Comment:
I don't feel super good about this for a few reasons:

* This check is still needed: https://github.com/chipsalliance/caliptra-sw/pull/377
* Ultimately we should probably also check that pub_key is a point on the curve

I agree that checking a single bool is bad for glitching purposes, but leaving it up to the caller to do the proper checks is also problematic. Two possible suggestions:

1. Still do all the checks here, but also return the R value for the caller to check again
2. Return a 32-bit magic number/status code denoting success. Then always compare against this value. I think this should mitigate most glitching related issues.

cc @mschilder123

Reply:
The R,S range check is definitely needed.

In both LMS and ECDSA verification the main vote comes from a comparison of large values (public key hash and computed R', respectively),  which offers intrinsic redundancy against single glitches. We do not want to compress that compare into a simple boolean but return codes in a sparse 32 bit space (maximized Hamming distances) are fairly resilient.

The caller still has to be careful (at the binary instruction stream level) when processing the single return value.

Perhaps there are better Rust patterns to add redundancy here?  A closure to be passed in which the verify library runs on success only, in addition to the 32b return code? Concerned callers could implement a redundant check in their closure.

Reply:
I like the idea of defining return codes with large hamming distances. Should only need two, pass/fail. Then the library can handle R-value comparison and the caller can just compare the returned value against 0xWhateverConstant.